aggregator/src/core.rs

```diff
@@ -638,118 +720,66 @@ pub(crate) fn conditional_constraints(
-                );
-                log::trace!(
-                    "candidate 3: {:?}",
-                    hash_input_len_cells[MAX_AGG_SNARKS * 2 + 5].value()
-                );
-
-                let mut data_hash_inputs_rec = rlc_config.mul(
-                    &mut region,
-                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 3],
-                    &flag1,
-                    &mut offset,
-                )?;
-                data_hash_inputs_rec = rlc_config.mul_add(
-                    &mut region,
-                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 4],
-                    &flag2,
-                    &data_hash_inputs_rec,
-                    &mut offset,
-                )?;
-                data_hash_inputs_rec = rlc_config.mul_add(
-                    &mut region,
-                    &hash_input_len_cells[MAX_AGG_SNARKS * 2 + 5],
-                    &flag3,
-                    &data_hash_inputs_rec,
-                    &mut offset,
-                )?;
-
-                // sanity check
-                assert_equal(&data_hash_inputs, &data_hash_inputs_rec);
-                region.constrain_equal(data_hash_inputs.cell(), data_hash_inputs_rec.cell())?;
                 log::trace!("rlc chip uses {} rows", offset);
-                Ok(num_of_valid_snarks_cell)
+                Ok(())
             },
         )
         .map_err(|e| Error::AssertionFailure(format!("aggregation: {e}")))?;
-    Ok(num_of_valid_snarks_cell[0].clone())
+    Ok(())
 }
 
+/// Input the all chunks' pi hash's data RLCs,
 /// generate a string of binary cells indicating
 /// if the i-th chunk is a valid chunk
-pub(crate) fn chunk_is_valid(
+///
+/// Assumption: the first chunk must be valid
+fn chunks_are_valid(
     rlc_config: &RlcConfig,
     region: &mut Region<Fr>,
-    num_of_valid_chunks: &AssignedCell<Fr, Fr>,
+    data_rlc_cells: &[AssignedCell<Fr, Fr>],
     offset: &mut usize,
-) -> Result<[AssignedCell<Fr, Fr>; MAX_AGG_SNARKS], halo2_proofs::plonk::Error> {
-    let mut res = vec![];
+) -> Result<[AssignedCell<Fr, Fr>; 10], halo2_proofs::plonk::Error> {
+    let chunk_pi_hash_rlc_cells = parse_pi_hash_rlc_cells(data_rlc_cells);
+
+    let one = rlc_config.load_private(region, &Fr::one(), offset)?;
+    let one_cell = rlc_config.one_cell(one.cell().region_index);
+    region.constrain_equal(one.cell(), one_cell)?;
 
-    for i in 0..MAX_AGG_SNARKS {
-        let value = rlc_config.load_private(region, &Fr::from(i as u64), offset)?;
-        let is_valid = rlc_config.is_smaller_than(region, &value, num_of_valid_chunks, offset)?;
-        res.push(is_valid);
+    let mut chunk_are_valid = vec![one];
+
+    for i in 1..MAX_AGG_SNARKS {
+        let is_padding = rlc_config.is_equal(
+            region,
+            chunk_pi_hash_rlc_cells[i - 1],
+            chunk_pi_hash_rlc_cells[i],
+            offset,
+        )?;
+        chunk_are_valid.push(rlc_config.not(region, &is_padding, offset)?);
     }
```

Comment:
Consider one corner case: 
>  [chunk0, chunk1, chunk2, chunk3, chunk3, chunk3, ..., chunk3] where chunk1 and chunk2 are empty chunk (i.e. their data hash = keccak("")).

Then the the `chunk_are_valid` that computed by the circuit would be `[true, true, false, true, false, false, ..., false]`. That is, only 3 out of 10 booleans are true which is not the case. Then honest prover will fail to produce a proof. 

To solve this, I think we can just use the following codes:
```rust
let is_padding = rlc_config.load_private(region, chunk_is_padding[i], offset)?;
rlc_config.conditional_enforce_equal(
    region, 
    chunk_pi_hash_rlc_cells[i-1], 
    chunk_pi_hash_rlc_cells[i], 
    offset,
)?;
```

Reply:
You're right, data_hash should be different even for empty chunks.